File: product-brain/src/product_brain/blocks/estimate.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from ..config import EstimateConfig
from ..models import Ticket, TicketRecord


@dataclass
class Reference:
    ticket: str
    title: str
    days: float
    loc: int
    files: int
    similarity: float


@dataclass
class Estimate:
    low: float
    high: float
    unit: str
    confidence: str
    references: list[Reference]


def _jaccard(a: set, b: set) -> float:
    if not a and not b:
        return 0.0
    return len(a & b) / max(len(a | b), 1)


def _ticket_files(rec: TicketRecord) -> set[str]:
    return {f.path for f in rec.files}


def similarity(
    target_files: set[str],
    target_symbols: set[str],
    target_labels: set[str],
    target_type: str,
    candidate: TicketRecord,
    candidate_labels: Optional[set[str]] = None,
) -> float:
    f = _jaccard(target_files, _ticket_files(candidate))
    s = _jaccard(target_symbols, set(candidate.symbols))
    label_score = _jaccard(target_labels, candidate_labels or set())
    type_score = 1.0 if target_type and target_type == candidate.type else 0.0
    return 0.55 * f + 0.20 * s + 0.15 * label_score + 0.10 * type_score


def estimate_effort(
    target: Ticket,
    candidates: list[TicketRecord],
    cfg: EstimateConfig,
    target_files: Optional[set[str]] = None,
    target_symbols: Optional[set[str]] = None,
    candidate_labels: Optional[dict[str, set[str]]] = None,
) -> Estimate:
    target_files = target_files or set()
    target_symbols = target_symbols or set()
    target_labels = set(target.labels)
    candidate_labels = candidate_labels or {}

    scored: list[tuple[float, TicketRecord]] = []
    for c in candidates:
        sim = similarity(
            target_files, target_symbols, target_labels, target.type,
            c, candidate_labels.get(c.ticket, set()),
        )
        if sim >= cfg.min_similarity:
            scored.append((sim, c))

    scored.sort(key=lambda x: x[0], reverse=True)
    top = scored[:5]

    refs = [
        Reference(
            ticket=c.ticket,
            title=c.title,
            days=(c.pr_open_to_merge_days if c.pr_open_to_merge_days is not None else c.duration_days),
            loc=c.loc_added + c.loc_removed,
            files=len(c.files),
            similarity=round(sim, 2),
        )
        for sim, c in top
    ]

    if not refs:
        return Estimate(low=0, high=0, unit=cfg.unit, confidence="low", references=[])

    weights = [r.similarity for r in refs]
    weighted_days = sum(r.days * w for r, w in zip(refs, weights)) / max(sum(weights), 1e-9)
    if len(refs) >= 2:
        mean = sum(r.days for r in refs) / len(refs)
        var = sum((r.days - mean) ** 2 for r in refs) / len(refs)
        sigma = math.sqrt(var)
    else:
        sigma = max(weighted_days * 0.3, 1.0)

    low = max(weighted_days - sigma, 0.5)
    high = weighted_days + sigma

    high_sim_count = sum(1 for r in refs if r.similarity >= 0.6)
    med_sim_count = sum(1 for r in refs if r.similarity >= 0.5)

    if high_sim_count >= cfg.min_references_for_high:
        confidence = "high"
    elif med_sim_count >= cfg.min_references_for_medium:
        confidence = "medium"
    else:
        confidence = "low"

    return Estimate(
        low=round(low, 1),
        high=round(high, 1),
        unit=cfg.unit,
        confidence=confidence,
        references=refs,
    )

File: product-brain-ts/src/blocks/estimate.ts
// Port target: ../product-brain/src/product_brain/blocks/estimate.py
import type { EstimateConfig } from "../config.js";
import type { Ticket, TicketRecord } from "../models.js";

export interface Reference {
  ticket: string;
  title: string;
  days: number;
  loc: number;
  files: number;
  similarity: number;
}

export interface Estimate {
  low: number;
  high: number;
  unit: string;
  confidence: "low" | "medium" | "high";
  references: Reference[];
}

function jaccard(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0;
  let inter = 0;
  for (const x of a) if (b.has(x)) inter += 1;
  const union = a.size + b.size - inter;
  return inter / Math.max(union, 1);
}

function ticketFiles(rec: TicketRecord): Set<string> {
  return new Set(rec.files.map((f) => f.path));
}

export function similarity(
  targetFiles: Set<string>,
  targetSymbols: Set<string>,
  targetLabels: Set<string>,
  targetType: string,
  candidate: TicketRecord,
  candidateLabels?: Set<string>,
): number {
  const f = jaccard(targetFiles, ticketFiles(candidate));
  const s = jaccard(targetSymbols, new Set(candidate.symbols));
  const l = jaccard(targetLabels, candidateLabels ?? new Set());
  const t = targetType && targetType === candidate.type ? 1 : 0;
  return 0.55 * f + 0.2 * s + 0.15 * l + 0.1 * t;
}

export function estimateEffort(
  target: Ticket,
  candidates: TicketRecord[],
  cfg: EstimateConfig,
  opts: {
    targetFiles?: Set<string>;
    targetSymbols?: Set<string>;
    candidateLabels?: Map<string, Set<string>>;
  } = {},
): Estimate {
  const targetFiles = opts.targetFiles ?? new Set<string>();
  const targetSymbols = opts.targetSymbols ?? new Set<string>();
  const targetLabels = new Set(target.labels);
  const candidateLabels = opts.candidateLabels ?? new Map();

  const scored: Array<[number, TicketRecord]> = [];
  for (const c of candidates) {
    const sim = similarity(
      targetFiles,
      targetSymbols,
      targetLabels,
      target.type,
      c,
      candidateLabels.get(c.ticket),
    );
    if (sim >= cfg.min_similarity) scored.push([sim, c]);
  }
  scored.sort((a, b) => b[0] - a[0]);
  const top = scored.slice(0, 5);

  const refs: Reference[] = top.map(([sim, c]) => ({
    ticket: c.ticket,
    title: c.title,
    days: c.prOpenToMergeDays ?? c.durationDays,
    loc: c.locAdded + c.locRemoved,
    files: c.files.length,
    similarity: Math.round(sim * 100) / 100,
  }));

  if (refs.length === 0) {
    return { low: 0, high: 0, unit: cfg.unit, confidence: "low", references: [] };
  }

  const weights = refs.map((r) => r.similarity);
  const wsum = Math.max(weights.reduce((a, b) => a + b, 0), 1e-9);
  const weightedDays = refs.reduce((a, r, i) => a + r.days * (weights[i] ?? 0), 0) / wsum;

  let sigma: number;
  if (refs.length >= 2) {
    const mean = refs.reduce((a, r) => a + r.days, 0) / refs.length;
    const variance = refs.reduce((a, r) => a + (r.days - mean) ** 2, 0) / refs.length;
    sigma = Math.sqrt(variance);
  } else {
    sigma = Math.max(weightedDays * 0.3, 1);
  }

  const low = Math.max(weightedDays - sigma, 0.5);
  const high = weightedDays + sigma;

  const highSim = refs.filter((r) => r.similarity >= 0.6).length;
  const medSim = refs.filter((r) => r.similarity >= 0.5).length;
  let confidence: "low" | "medium" | "high" = "low";
  if (highSim >= cfg.min_references_for_high) confidence = "high";
  else if (medSim >= cfg.min_references_for_medium) confidence = "medium";

  return {
    low: Math.round(low * 10) / 10,
    high: Math.round(high * 10) / 10,
    unit: cfg.unit,
    confidence,
    references: refs,
  };
}
